kpartx/kpartx.8
.\" ----------------------------------------------------------------------------
.\" Update the date below if you make any significant change.
.\" ----------------------------------------------------------------------------
.
.TH KPARTX 8 2019-04-27 Linux
.
.
.\" ----------------------------------------------------------------------------
.SH NAME
.\" ----------------------------------------------------------------------------
.
kpartx \- Create device maps from partition tables.
.
.
.\" ----------------------------------------------------------------------------
.SH SYNOPSIS
.\" ----------------------------------------------------------------------------
.
.B kpartx
.RB [\| \-a | \-d | \-u | \-l \|]
.RB [\| \-r \|]
.RB [\| \-p \|]
.RB [\| \-f \|]
.RB [\| \-g \|]
.RB [\| \-s | \-n \|]
.RB [\| \-v \|]
.B wholedisk
.
.
.\" ----------------------------------------------------------------------------
.SH DESCRIPTION
.\" ----------------------------------------------------------------------------
.
This tool, derived from util-linux' partx, reads partition tables on specified
device and create device maps over partitions segments detected. It is called
from hotplug upon device maps creation and deletion.
.
.
.\" ----------------------------------------------------------------------------
.SH OPTIONS
.\" ----------------------------------------------------------------------------
.
.TP
.B \-a
Add partition mappings.
.
.TP
.B \-d
Delete partition mappings.
.
.TP
.B \-u
Update partition mappings.
.
.TP
.B \-l
List partition mappings that would be added \-a.
.
.TP
.B \-r
Read-only partition mappings.
.
.TP
.B \-p
Set device name-partition number delimiter.
.
.TP
.B \-f
Force creation of mappings; overrides 'no_partitions' feature.
.
.TP
.B \-g
Force GUID partition table (GPT).
.
.TP
.B \-s
Sync mode (Default). Don't return until the partitions are created.
.
.TP
.B \-n
Nosync mode. Return before the partitions are created.
.
.TP
.B \-v
Operate verbosely.
.
.
.\" ----------------------------------------------------------------------------
.SH EXAMPLE
.\" ----------------------------------------------------------------------------
.
To mount all the partitions in a raw disk image:
.IP
kpartx \-av disk.img
.PP
This will output lines such as:
.IP
add map loop1p1 (254:4): 0 409597 linear 7:1 3
.PP
The \fIloop1p1\fR is the name of a device file under \fI/dev/mapper\fR which you
can use to access the partition, for example to fsck it:
.IP
fsck /dev/mapper/loop1p1
.PP
When you're done, you need to remove the devices:
.IP
kpartx \-d disk.img
.
.
.\" ----------------------------------------------------------------------------
.SH "SEE ALSO"
.\" ----------------------------------------------------------------------------
.
.BR multipath (8)
.BR multipathd (8)
.BR hotplug (8)
.
.
.\" ----------------------------------------------------------------------------
.SH AUTHORS
.\" ----------------------------------------------------------------------------
.
This man page was assembled By Patrick Caulfield for the Debian project.
.PP
\fImultipath-tools\fR was developed by Christophe Varoqui <christophe.varoqui@opensvc.com>
and others.
.\" EOF


NEWS.md
# multipath-tools Release Notes

## Stable branches and new versioning scheme

Beginning with 0.11, the second digit in the multipath-tools version will be
used for upstream "major" releases. The 3rd and last digit will denote stable
releases in the future, containing only bug fixes on top of the last major
release. These bug fixes will be tracked in stable branches.

See [README.md](README.md) for additional information.

## multipath-tools 0.14.4, 2026/07

### User-visible changes

* Fix ALUA asymmetric access state descriptions in multipathd logs, so that
  the same terms are used as by the kernel ("lba-dependent", "transitioning").
* Don't set a hardware handler for bio-based multipath devices. The kernel
  rejects this anyway.

### Bug fixes

* Fix WWID detection for legacy devices that use the older SCSI-2 VPD page
  0x83 format for their device identifier.
* kpartx: Fix an integer overflow in the GPT partition table size calculation.
  A crafted partition table with an extremely large number of partition entries
  could trigger the overflow.
* kpartx: Fix several issues in the DASD partition table reader that could be
  triggered by a maliciously crafted disk image.
* Fix duplicate "checker timed out" log messages when `log_checker_err` is
  set to `once`. Fixes 0.5.0.
* Avoid potential buffer overflows in the iet and datacore prioritizers.
  Commit 4611f97.
* iet prioritizer: avoid misleading error message with systemd 256 and
  newer, and properly use udev to derive path parameters.
  Fixes [#145](https://github.com/opensvc/multipath-tools/issues/145).
* Even with the libudev wrapper code introduced in 0.14.0, multipathd ran into
  use-after-free errors in tests where multipathd was restarted frequently.
  Fix this by preventing thread cancellation during libudev calls.
  Fixes [#152](https://github.com/opensvc/multipath-tools/issues/152).
* libmpathpersist: Fix self-preemption without holding a reservation.
  Fixes 0.13.0.

### Other changes

* Man page improvements

### CI

* Removed the `check_spelling` GitHub Action after a
  [security incident](https://github.com/jsoref/2026-06-16-credential-leak)
  reported against the action's source repository.
* Added coredump collection in the CI.
* GitHub action updates related to node 20 depreciation on GitHub.
* Fixes for various errors in the CI.

## multipath-tools 0.14.3, 2026/02

* Fix boot failures with multipath introduced in 0.14.1. Commit 01cc89c.

## multipath-tools 0.14.2, 2026/01

### Bug fixes

* Fix a possible NULL pointer dereference in the path purging code.
  Fixes 0.14.0. Commit 86ff48b.

### CI

* Updated the `test_kpartx` test script, and added it to the
  `basic-build-and-ci` workflow.

## multipath-tools 0.14.1, 2026/01

### Bug fixes

* kpartx: Fix freeing static buffer when operating on regular files.
  Fixes 0.14.0. Commit fab5d44.
  Fixes [#139](https://github.com/opensvc/multipath-tools/issues/139).
* Fix initialization of paths that were offline during path detection.
  Commit 1942fb1.
* Fix printing the "path offline" log message for offline paths that don't
  have a path checker configured. Commit 1a364a1.

### Other changes

* If path devices that are members of multipath maps aren't detected during
  multipathd startup or `reconfigure`, don't add them to newly created maps
  in the first place. In previous versions, such paths would be added to the
  maps, only to be removed later. Commit a04be55.
* Improve the detection of "busy" state in the `show status` command, such
  that reading udev events from udevd is also counted as busy.
  Commit 7fdd93b.

## multipath-tools 0.14.0, 2026/01

### User-visible changes

#### Automatic removal of disconnected SCSI devices

This release adds support for automatically purging SCSI devices that become
disconnected at the storage target (e.g. because a LUN became unmapped by an
administrator). It's a long-standing problem that such path devices become
"stale" on the Linux side. When they are re-mapped on the storage side,
possibly with different parameters, the kernel may not correctly detect
the changes. By removing such stale devices, the kernel will be forced to
re-probe them, which fixes this problem.

The feature is controlled by a new configuration option `purge_disconnected`
that can be set in `multipath.conf` per multipath device, hardware entry, or
globally. The feature is switched off by default.

When `purge_disconnected` is enabled, SCSI devices for which the storage array
reports a "LUN NOT SUPPORTED" status are removed from the system, preventing
stale device entries from accumulating.

As this is a new feature, users are advised to test it in their environment
before enabling it in production.

### Bug fixes

* Make sure multipathd registers keys all paths of a multipath map after
  mpathpersist registered a key for a map. multipathd could fail to do so in
  some cases, e.g. if paths become unavailable or available while the
  registration was taking place. Fixes 0.13.0. Commits f7d6cd1, 4f3036b.
* Fix `mpathpersist --report-capabilities` output. Fixes 0.5.0. Commit
  c8ed5e6.
* Improve error handling when retrying REGISTER AND IGNORE persistent
  reservations commands. Fixes 0.13.0. Commit c971036.
* Fix command descriptions in the multipathd man page. Fixes 0.9.2.
  Commit f3ba2e7.
* Fix an undefined symbol error with the LLVM lld linker. Commit a298603.
  Fixes [#132](https://github.com/opensvc/multipath-tools/issues/132), 0.10.0.
* Fix ISO C23 compatibility issue causing errors with new compilers.
  Commit 9f611e2.
* Fix use-after-free error in free_pgvec(). Commit 7a07023.
  Fixes [#128](https://github.com/opensvc/multipath-tools/issues/128), 0.12.0.
* Fix memory leak caused by not joining the "init unwinder" thread.
  Fixes 0.8.6. Commit 29f262b.
* Fix memory leaks in kpartx. Commits 8c39e60, 02e0933. Fixes any version.
* Fix a particular strict aliasing bug in the bitfield code.

kpartx/kpartx.c
/*
 * Source: copy of util-linux' partx partx.c
 *
 * Copyrights of the original file applies
 * Copyright (c) 2004, 2005 Christophe Varoqui
 * Copyright (c) 2005 Kiyoshi Ueda
 * Copyright (c) 2005 Lars Soltau
 */

/*
 * Given a block device and a partition table type,
 * try to parse the partition table, and list the
 * contents. Optionally add or remove partitions.
 *
 * Read wholedisk and add all partitions:
 *	kpartx [-a|-d|-l] [-v] wholedisk
 *
 * aeb, 2000-03-21
 * cva, 2002-10-26
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <stdint.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/sysmacros.h>
#include <sys/types.h>
#include <ctype.h>
#include <libdevmapper.h>

#include "autoconfig.h"
#include "devmapper.h"
#include "crc32.h"
#include "lopart.h"
#include "kpartx.h"
#include "version.h"

#define SIZE(a) (sizeof(a)/sizeof((a)[0]))

#define MAXTYPES	64
#define MAXSLICES	256
#define DM_TARGET	"linear"
#define LO_NAME_SIZE    64
#define PARTNAME_SIZE	128
#define DELIM_SIZE	8

struct slice slices[MAXSLICES];

enum action { LIST, ADD, DELETE, UPDATE };

struct pt {
	char *type;
	ptreader *fn;
} pts[MAXTYPES];

int ptct = 0;
int udev_sync = 1;

static void
addpts(char *t, ptreader f)
{
	if (ptct >= MAXTYPES) {
		fprintf(stderr, "addpts: too many types\n");
		exit(1);
	}
	pts[ptct].type = t;
	pts[ptct].fn = f;
	ptct++;
}

static void
initpts(void)
{
	addpts("gpt", read_gpt_pt);
	addpts("dos", read_dos_pt);
	addpts("bsd", read_bsd_pt);
	addpts("solaris", read_solaris_pt);
	addpts("unixware", read_unixware_pt);
	addpts("dasd", read_dasd_pt);
	addpts("mac", read_mac_pt);
	addpts("sun", read_sun_pt);
	addpts("ps3", read_ps3_pt);
}

static char short_opts[] = "rladfgvp:t:snu";

/* Used in gpt.c */
int force_gpt=0;

int force_devmap=0;

static int
usage(void) {
	printf(VERSION_STRING);
	printf("Usage:\n");
	printf("  kpartx [-a|-d|-u|-l] [-r] [-p] [-f] [-g] [-s|-n] [-v] wholedisk\n");
	printf("\t-a add partition devmappings\n");
	printf("\t-r devmappings will be readonly\n");
	printf("\t-d del partition devmappings\n");
	printf("\t-u update partition devmappings\n");
	printf("\t-l list partitions devmappings that would be added by -a\n");
	printf("\t-p set device name-partition number delimiter\n");
	printf("\t-g force GUID partition table (GPT)\n");
	printf("\t-f force devmap create\n");
	printf("\t-v verbose\n");
	printf("\t-n nosync mode. Return before the partitions are created\n");
	printf("\t-s sync mode (Default). Don't return until the partitions are created\n");
	return 1;
}

static void
set_delimiter (char * device, char * delimiter)
{
	char * p = device;

	if (*p == 0x0)
		return;

	while (*(++p) != 0x0)
		continue;

	if (isdigit(*(p - 1)))
		*delimiter = 'p';
}

static int
find_devname_offset (char * device)
{
	char *p, *q;

	q = p = device;

	while (*p) {
		if (*p == '/')
			q = p + 1;
		p++;
	}

	return (int)(q - device);
}

static int
check_uuid(char *uuid, char *part_uuid, char **err_msg) {
	char *map_uuid = strchr(part_uuid, '-');
	if (!map_uuid || strncmp(part_uuid, "part", 4) != 0) {
		*err_msg = "not a kpartx partition";
		return -1;
	}
	map_uuid++;
	if (strcmp(uuid, map_uuid) != 0) {
		*err_msg = "a partition of a different device";
		return -1;
	}
	return 0;
}

static void *
xmalloc (size_t size) {
	void *t;

	if (size == 0)
		return NULL;

	t = malloc (size);

	if (t == NULL) {
		fprintf(stderr, "Out of memory\n");
		exit(1);
	}

	return t;
}

static void cleanup_charp(char **p)
{
	free(*p);
}

int
main(int argc, char **argv){
	int i, j, m, n, op, off, arg, c, d, ro=0;
	int fd = -1;
	struct slice all;
	struct pt *ptp;
	enum action what = LIST;
	char *type, *diskdevice, *device;
	int verbose = 0;
	char partname[PARTNAME_SIZE], params[PARTNAME_SIZE + 16];
	char *loopdev __attribute__((cleanup(cleanup_charp))) = NULL;
	char *delim __attribute__((cleanup(cleanup_charp))) = NULL;
	char *uuid __attribute__((cleanup(cleanup_charp))) = NULL;
	char *_mapname __attribute__((cleanup(cleanup_charp))) = NULL;
	char *mapname;
	int loopcreated = 0;
	struct stat buf;

	initpts();
	init_crc32();

	type = device = diskdevice = NULL;
	memset(&all, 0, sizeof(all));
	memset(&partname, 0, sizeof(partname));

	if (argc < 2) {
		usage();
		exit(1);
	}

	while ((arg = getopt(argc, argv, short_opts)) != EOF)
		switch(arg) {
		case 'r':
			ro=1;
			break;
		case 'f':
			force_devmap=1;
			break;
		case 'g':
			force_gpt=1;
			break;
		case 't':
			type = optarg;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'p':
			delim = strdup(optarg);
			if (!delim)
				exit(1);
			break;
		case 'l':
			what = LIST;
			break;
		case 'a':
			what = ADD;
			break;
		case 'd':
			what = DELETE;
			break;
		case 's':
			udev_sync = 1;
			break;
		case 'n':
			udev_sync = 0;
			break;
		case 'u':
			what = UPDATE;
			break;
		default:
			usage();
			exit(1);
		}

#ifdef LIBDM_API_COOKIE
	if (!udev_sync)
		dm_udev_set_sync_support(0);
	else
		dm_udev_set_sync_support(1);
#endif

	if (dm_prereq(DM_TARGET, 0, 0, 0) && (what == ADD || what == DELETE || what == UPDATE)) {
		fprintf(stderr, "device mapper prerequisites not met\n");
		exit(1);
	}

	if (optind == argc - 2) {
		device = argv[optind];
		diskdevice = argv[optind+1];
	} else if (optind == argc - 1) {
		diskdevice = device = argv[optind];
	} else {
		usage();
		exit(1);
	}

	if (stat(device, &buf)) {
		printf("failed to stat() %s\n", device);
		exit (1);
	}

	if (S_ISREG (buf.st_mode)) {
		/* already looped file ? */
		char rpath[PATH_MAX];
		if (realpath(device, rpath) == NULL) {
			fprintf(stderr, "Error: %s: %s\n", device,
				strerror(errno));
			exit (1);
		}
		loopdev = find_loop_by_file(rpath);

		if (!loopdev && what == DELETE)
			exit (0);

		if (!loopdev) {
			if (set_loop(&loopdev, rpath, 0, &ro)) {
				fprintf(stderr, "can't set up loop\n");
				exit (1);
			}
			loopcreated = 1;
		}
		device = loopdev;

		if (stat(device, &buf)) {
			printf("failed to stat() %s\n", device);
			exit (1);
		}
	}
	else if (!S_ISBLK(buf.st_mode)) {
		fprintf(stderr, "invalid device: %s\n", device);
		exit(1);
	}

	off = find_devname_offset(device);

	if (!loopdev) {
		_mapname = dm_mapname(major(buf.st_rdev), minor(buf.st_rdev));
		if (_mapname)
			uuid = dm_mapuuid(_mapname);
	}

	mapname = _mapname;
	if (!mapname)
		mapname = device + off;

	/*
	 * We are called for a non-DM device.
	 * Make up a fake UUID for the device, unless "-d -f" is given.
	 * This allows deletion of partitions created with older kpartx
	 * versions which didn't use the fake UUID during creation.
	 */
	if (!uuid && !(what == DELETE && force_devmap)) {
		uuid = nondm_create_uuid(buf.st_rdev);
		if (!uuid)
			exit(1);
	}

	if (delim == NULL) {
		delim = xmalloc(DELIM_SIZE);
		memset(delim, 0, DELIM_SIZE);
		set_delimiter(mapname, delim);
	}

	/* add/remove partitions to the kernel devmapper tables */
	int r = 0;

	if (what == DELETE) {
		r = dm_remove_partmaps(mapname, uuid, buf.st_rdev,
				       verbose);
		if (loopdev) {
			if (del_loop(loopdev)) {
				if (verbose)
					fprintf(stderr, "can't del loop : %s\n",
					       loopdev);
				r = 1;
			} else if (verbose)
				fprintf(stderr, "loop deleted : %s\n", loopdev);
		}
		goto end;
	}

	fd = open(device, O_RDONLY | O_DIRECT);

	if (fd == -1) {
		perror(device);
		exit(1);
	}

	for (i = 0; i < ptct; i++) {
		ptp = &pts[i];

		if (type && strcmp(type, ptp->type))
			continue;

		/* here we get partitions */
		n = ptp->fn(fd, all, slices, SIZE(slices));

#ifdef DEBUG
		if (n >= 0)
			printf("%s: %d slices\n", ptp->type, n);
#endif
		if (n <= 0)
			continue;

		switch(what) {
		case LIST:
			for (j = 0, c = 0, m = 0; j < n; j++) {
				if (slices[j].size == 0)
					continue;
				if (slices[j].container > 0) {
					c++;
					continue;
				}

				slices[j].minor = m++;

				printf("%s%s%d : 0 %" PRIu64 " %s %" PRIu64"\n",
				       mapname, delim, j+1,
				       slices[j].size, device,
				       slices[j].start);
			}
			/* Loop to resolve contained slices */
			d = c;
			while (c) {
				for (j = 0; j < n; j++) {
					uint64_t start;
					int k = slices[j].container - 1;

					if (slices[j].size == 0)
						continue;
					if (slices[j].minor > 0)
						continue;
					if (slices[j].container == 0)
						continue;
					slices[j].minor = m++;

					start = slices[j].start - slices[k].start;
					printf("%s%s%d : 0 %" PRIu64 " /dev/dm-%d %" PRIu64 "\n",
					       mapname, delim, j+1,
					       slices[j].size,
					       slices[k].minor, start);
					c--;
				}
				/* Terminate loop if nothing more to resolve */
				if (d == c)
					break;
			}

			break;

		case ADD:
		case UPDATE:
			/* ADD and UPDATE share the same code that adds new partitions. */
			for (j = 0, c = 0; j < n; j++) {
				char *part_uuid = NULL, *reason;

				if (slices[j].size == 0)
					continue;

				/* Skip all contained slices */
				if (slices[j].container > 0) {
					c++;
					continue;
				}

				if (safe_sprintf(params, "%d:%d %" PRIu64 ,
						 major(buf.st_rdev), minor(buf.st_rdev), slices[j].start)) {
					fprintf(stderr, "params too small\n");
					exit(1);
				}

				op = dm_find_part(mapname, delim, j + 1, uuid,
						  partname, sizeof(partname),
						  &part_uuid, verbose);
				if (op == DFP_ERR) {
					r++;
					continue;
				}

				if (part_uuid && uuid) {
					if (check_uuid(uuid, part_uuid, &reason) != 0) {
						fprintf(stderr, "%s is already in use, and %s\n", partname, reason);
						r++;
						free(part_uuid);
						continue;
					}
					free(part_uuid);
				}

				if (!dm_addmap(op, partname, DM_TARGET, params,
					       slices[j].size, ro, uuid, j+1,
					       buf.st_mode & 0777, buf.st_uid,
					       buf.st_gid)) {
					fprintf(stderr, "create/reload failed on %s\n",
						partname);
					r++;
					continue;
				}
				if (op == DM_DEVICE_RELOAD &&
				    !dm_simplecmd(DM_DEVICE_RESUME, partname,
						  1, MPATH_UDEV_RELOAD_FLAG)) {
					fprintf(stderr, "resume failed on %s\n",
						partname);
					r++;
					continue;
				}

				dm_devn(partname, &slices[j].major,
					&slices[j].minor);

				if (verbose)
					printf("add map %s (%d:%d): 0 %" PRIu64 " %s %s\n",
					       partname, slices[j].major,
					       slices[j].minor, slices[j].size,
					       DM_TARGET, params);
			}
			/* Loop to resolve contained slices */
			d = c;
			while (c) {
				for (j = 0; j < n; j++) {
					char *part_uuid = NULL, *reason;
					int k = slices[j].container - 1;

					if (slices[j].size == 0)
						continue;

					/* Skip all existing slices */
					if (slices[j].minor > 0)
						continue;

					/* Skip all simple slices */
					if (slices[j].container == 0)
						continue;

					/* Check container slice */
					if (slices[k].size == 0)
						fprintf(stderr, "Invalid slice %d\n",
							k);

					if (safe_sprintf(params, "%d:%d %" PRIu64,
							 major(buf.st_rdev), minor(buf.st_rdev),
							 slices[j].start)) {
						fprintf(stderr, "params too small\n");
						exit(1);
					}

					op = dm_find_part(mapname, delim,
							  j + 1, uuid, partname,
							  sizeof(partname),
							  &part_uuid, verbose);
					if (op == DFP_ERR) {
						r++;
						continue;
					}

					if (part_uuid && uuid) {
						if (check_uuid(uuid, part_uuid, &reason) != 0) {
							fprintf(stderr, "%s is already in use, and %s\n", partname, reason);
							free(part_uuid);
							continue;
						}
						free(part_uuid);
					}

					dm_addmap(op, partname, DM_TARGET, params,
						  slices[j].size, ro, uuid, j+1,
						  buf.st_mode & 0777,
						  buf.st_uid, buf.st_gid);

					if (op == DM_DEVICE_RELOAD)
						dm_simplecmd(DM_DEVICE_RESUME,
							     partname, 1,
							     MPATH_UDEV_RELOAD_FLAG);
					dm_devn(partname, &slices[j].major,
						&slices[j].minor);

					if (verbose)
						printf("add map %s (%d:%d): 0 %" PRIu64 " %s %s\n",
						       partname, slices[j].major, slices[j].minor, slices[j].size,
						       DM_TARGET, params);
					c--;
				}
				/* Terminate loop */
				if (d == c)
					break;
			}

			if (what == ADD) {
				/* Skip code that removes devmappings for deleted partitions */
				break;
			}

			for (j = MAXSLICES-1; j >= 0; j--) {
				char *part_uuid = NULL, *reason;

				if (slices[j].size ||
				    dm_find_part(mapname, delim, j + 1, uuid, partname,
						 sizeof(partname), &part_uuid,
						 verbose) != DFP_DEVICE_RELOAD)
					continue;

				if (part_uuid && uuid) {
					if (check_uuid(uuid, part_uuid, &reason) != 0) {
						fprintf(stderr, "%s is %s. Not removing\n", partname, reason);
						free(part_uuid);
						continue;
					}
					free(part_uuid);
				}

				if (!dm_simplecmd(DM_DEVICE_REMOVE,
						  partname, 1, 0)) {
					fprintf(stderr, "failed to remove %s",
						partname);
					r++;
					continue;
				}
				if (verbose)
					printf("del devmap : %s\n", partname);
			}

		default:
			break;

		}
		if (n > 0)
			break;
	}
	if (fd != -1)
		close(fd);
	if (what == LIST && loopcreated) {
		if (del_loop(device)) {
			if (verbose)
				fprintf(stderr, "can't del loop : %s\n",
					device);
			exit(1);
		}
		if (verbose)
			fprintf(stderr, "loop deleted : %s\n", device);
	}
end:
	dm_lib_exit();

	return r;
}

/*
 * sseek: seek to specified sector
 */

static int
sseek(int fd, unsigned int secnr, int secsz) {
	off64_t in, out;
	in = ((off64_t) secnr * secsz);
	out = 1;

	if ((out = lseek64(fd, in, SEEK_SET)) != in)
	{
		fprintf(stderr, "llseek error\n");
		return -1;
	}
	return 0;
}

int
aligned_malloc(void **mem_p, size_t align, size_t *size_p)
{
	static size_t pgsize = 0;
	size_t size;
	int err;

	if (!mem_p || !align || (size_p && !*size_p))
		return EINVAL;

	if (!pgsize)
		pgsize = getpagesize();

	if (size_p)
		size = ((*size_p + align - 1) / align) * align;
	else
		size = pgsize;

	err = posix_memalign(mem_p, pgsize, size);
	if (!err && size_p)
		*size_p = size;
	return err;
}

/* always in sector size blocks */
static
struct block {
	unsigned int secnr;
	char *block;
	struct block *next;
} *blockhead;

/* blknr is always in 512 byte blocks */
char *
getblock (int fd, unsigned int blknr) {
	int secsz = get_sector_size(fd);
	unsigned int blks_per_sec = secsz / 512;
	unsigned int secnr = blknr / blks_per_sec;
	unsigned int blk_off = (blknr % blks_per_sec) * 512;
	struct block *bp;

	for (bp = blockhead; bp; bp = bp->next)

		if (bp->secnr == secnr)
			return bp->block + blk_off;

	if (sseek(fd, secnr, secsz))
		return NULL;

	bp = xmalloc(sizeof(struct block));
	bp->secnr = secnr;
	bp->next = blockhead;
	blockhead = bp;
	if (aligned_malloc((void **)&bp->block, secsz, NULL)) {
		fprintf(stderr, "aligned_malloc failed\n");
		exit(1);
	}

	if (read(fd, bp->block, secsz) != secsz) {
		fprintf(stderr, "read error, sector %d\n", secnr);
		blockhead = bp->next;
		free(bp->block);
		free(bp);
		return NULL;
	}

	return bp->block + blk_off;
}

int
get_sector_size(int filedes)
{
	int rc, sector_size = 512;

	rc = ioctl(filedes, BLKSSZGET, &sector_size);
	if (rc)
		sector_size = 512;
	return sector_size;
}
